File: 55.jump-game.py
#
# @lc app=leetcode id=55 lang=python3
#
# [55] Jump Game
#

# Time:  O(n)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        reach = 0
        for i, jump in enumerate(nums):
            if i > reach:
                break
            reach = max(reach, i + jump)
        return reach >= len(nums) - 1


# @lc code=end

File: 55.jump-game.js
/*
 * @lc app=leetcode id=55 lang=javascript
 *
 * [55] Jump Game
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function (nums) {
  let reach = 0;
  for (let i = 0; i < nums.length; ++i) {
    if (i > reach) {
      return false;
    }
    reach = Math.max(reach, i + nums[i]);
  }
  return reach >= nums.length - 1;
};
// @lc code=end
